### PR TITLE
read logger topics from a topic list file

### PR DESCRIPTION
**Pull request by kd0aij**
tested on Pixracer

@bkueng @LorenzMeier We could use either a new parameter or a startup variable to specify the name of the topic list file. It seems like a parameter would be the most flexible way to do this.

Advice and/or criticism appreciated.

### CODE REVIEW

File: src/modules/logger/logger.cpp [hunk 1]
```diff
@@ -411,6 +411,59 @@ bool Logger::copy_if_updated_multi(LoggerSubscription &sub, int multi_instance,
 	return updated;
 }
 
+int Logger::add_topics_from_file(const char *fname)
+{
+	FILE		*fp;
+	char		line[80];
+	char		topic_name[40];
```
Comment (bkueng) [inline src/modules/logger/logger.cpp]: making `topic_name` smaller than line, can lead to an overflow in sscanf. It's unlikely, but you never know what a user might enter.

File: src/modules/logger/logger.cpp [hunk 2]
```diff
@@ -411,6 +411,59 @@ bool Logger::copy_if_updated_multi(LoggerSubscription &sub, int multi_instance,
...
+
+		/* get a line, bail on error/EOF */
+		line[0] = '\0';
+
+		if (fgets(line, sizeof(line), fp) == NULL) {
+			break;
+		}
+
+		/* skip comment lines */
+		if ((strlen(line) < 2) || (line[1] == '#')) {
```
Comment (bkueng) [inline src/modules/logger/logger.cpp]: should be `line[0]`

File: src/modules/logger/logger.cpp [hunk 3]
```diff
@@ -411,6 +411,59 @@ bool Logger::copy_if_updated_multi(LoggerSubscription &sub, int multi_instance,
...
+			continue;
+		}
+
+		int nfields = sscanf(line, "%s, %u", &topic_name[0], &interval);
```
Comment (bkueng) [inline src/modules/logger/logger.cpp]: `&topic_name[0]` --> use `topic_name` (below as well)

**Comment by bkueng** [comment]
If I see that right, putting a file in `px4fmu_common` requires a separate file in `posix_configs` to run SITL. We should make sure to have only one default topics file.

**Comment by kd0aij** [comment]
@bkueng Thanks for pointing out those problems. What would be the best way to provide the defaults? One option would be to use hard-coded defaults in logger.cpp if no file is present on the SD card.

**Comment by bkueng** [comment]
I'd leave them in logger.cpp. If you have a file in `px4fmu_common` and change it, you need to recompile too, so I don't see an advantage in having a separate file.

**Comment by kd0aij** [comment]
rebased on master

**Comment by kd0aij** [comment]
This is functional now and tested on fmu-v4. The default topics are subscribed in function add_default_topics in logger.cpp. That function is invoked if reading and parsing a file named /etc/logging/topic_list.txt on the SD card fails to add any topics.

**Comment by kd0aij** [comment]
@tumbili The default topics are currently just the ones I've been using to do impulse response analysis, and probably need additions and deletions to be optimal. The default topics could also be made dependent on things like the parameter EKF2_REC_RPL, or we could just rely on the user putting the appropriate topic_list file on the SD card in addition to setting the parameter. Comments?

File: ROMFS/px4fmu_common/logging/default_topics.txt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+vehicle_local_position_setpoint, 200
+rc_channels, 20
+#ekf2_innovations, 20
+commander_state, 100
+vehicle_local_position, 200
+vehicle_global_position, 200
+system_power, 100
+servorail_status, 200
+mc_att_ctrl_status, 50
+#control_state
```
Comment (LorenzMeier) [inline ROMFS/px4fmu_common/logging/default_topics.txt]: Is this file necessary? Or should this rather go into docs?

**Comment by kd0aij** [comment]
@LorenzMeier I just forgot to delete the default topics file after I moved them back into logger.cpp.

**Comment by bkueng** [comment]
@kd0aij sorry, I lost track of this. Do you think `on/off` is needed? You get a very similar behavior with `logger start -e` `logger stop`.
Then in future please use apporpriate log levels (`PX4_INFO/WARN/ERR/DEBUG`), otherwise looks good.

**Comment by kd0aij** [comment]
how do you set the log level? The default is no debug output. Looks like
it's a compile-time only option, but nobody has answered my questions about
it.
I added on/off because sdlog2 had it and commander was refusing to arm.
also, start/stop doesn't turn logging on or off

**Comment by bkueng** [comment]
All levels are enabled, except for debug. I just tried to enable the debug system-wide but it renders it unusable. What you can do is to add `-DDEBUG_BUILD` to compile flags in your module's CMakeLists.txt file:
```
	COMPILE_FLAGS -Os -DDEBUG_BUILD
```

`logger start -e` turns logging on immediately w/o requiring to arm.

**Comment by kd0aij** [comment]
Thanks. I thought I tried the compiler switch, but probably put it in the
wrong place.
So start/stop will create a new log directory each time vs. on/off which
will keep things in the same "session"?

**Comment by bkueng** [comment]
ok, fair enough. rebased & applied
